### PR TITLE
[Feature Enhanced] Generalize F1-measure to FBeta-measure

### PR DESCRIPTION
**Pull request by WrRan**
Inspired by [`precision_recall_fscore_support`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/classification.py#L1095) in `sklearn`, I generalize F1-measure to FBeta-measure.

### CODE REVIEW

File: allennlp/training/metrics/fbeta_measure.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        f1-measures : List[float]
+        """
+        if self._tp_sum is None:
+            return [], [], []
+
+        tp_sum = self._tp_sum
+        pred_sum = self._pred_sum
+        true_sum = self._true_sum
+
+        if self._average == 'micro':
```
Comment (WrRan) [inline allennlp/training/metrics/fbeta_measure.py]: `mypy` complains:
```bash

[15:48:40] :	 [Step 5/10] allennlp/training/metrics/fbeta_measure.py:154: error: None has no attribute "sum"
[15:48:40] :	 [Step 5/10] allennlp/training/metrics/fbeta_measure.py:155: error: None has no attribute "sum"
[15:48:40] :	 [Step 5/10] allennlp/training/metrics/fbeta_measure.py:156: error: None has no attribute "sum"
```
But I do not know how to fix it. =(

Comment (WrRan) [reply allennlp/training/metrics/fbeta_measure.py]: I found a solution, but maybe it isn't a good idea.

**Comment by WrRan** [comment]
Please have a look. :)

File: allennlp/tests/training/metrics/fbeta_measure_test.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        self.assertRaises(ConfigurationError, FBetaMeasure,
+                          beta=0.0)
+
+        # Bad average option
+        self.assertRaises(ConfigurationError, FBetaMeasure,
+                          average='mega')
+
+    def test_runtime_errors(self):
+        fbeta = FBetaMeasure()
+        # never call metric
```
[the remark this answers is not in the document]
Comment (WrRan) [reply allennlp/tests/training/metrics/fbeta_measure_test.py]: The subtle point is that if we don't call this `metric`, we don't know how many classes are in the data, and there is no way to return a tensor of the appropriate shape, i.e. `None` or a tensor of shape `(num_classes, )`. Thus, I think when we use `FbetaMeasure` directly and specify `average=None`, throwing an exception might be a better idea if this metric is never called.

As for `F1Measure`, we can handle it as before due to backward compatibility.

Does it make sense?

[the remark this answers is not in the document]
Comment (brendan-ai2) [reply allennlp/tests/training/metrics/fbeta_measure_test.py]: Yes, that seems reasonable. Thank you for explaining.

**Comment by WrRan** [comment]
It failed due to the following error:
```txt
undocumented module: allennlp.training.metrics.fbeta_measure
```
How can I fix it? :frowning:

**Comment by matt-gardner** [comment]
@WrRan, that failure means you need to add your module to our API documentation, [here](https://github.com/allenai/allennlp/tree/master/doc/api).  Let us know if you need help figuring out exactly how to do that.

**Comment by brendan-ai2** [comment]
LGTM. @WrRan, thanks for following up on everything! Please feel free to ping us if a PR isn't getting attention. It's much easier to tell that a PR has been updated when there are explicit replies to the earlier comments or an explicit request for "another look".

**Comment by huntzhan** [comment]
Hi @WrRan, there's an issue if we pass predictions with bach size = 1.

```
>       pred_bins = argmax_predictions[mask].long()
E       IndexError: too many indices for tensor of dimension 0

(Pdb) mask
tensor([1], dtype=torch.uint8)
(Pdb) argmax_predictions
tensor(1.)
(Pdb) predictions
tensor([[0.2862, 0.3479, 0.1627, 0.2033]])
```

And it's caused by 

https://github.com/allenai/allennlp/blob/95b439da692eff28532179aaa1dbedc64ebf658a/allennlp/training/metrics/fbeta_measure.py#L125

I think we don't need to squeeze the `argmax_predictions`.

**Comment by brendan-ai2** [comment]
Thanks for finding this, @huntzhan! I've made a fix here: https://github.com/allenai/allennlp/pull/2777

**Pull request merged**